torch/_inductor/codegen/triton_foreach.py

```diff
@@ -1,81 +1,140 @@
+from ..utils import ceildiv
 from ..virtualized import V
 from .common import IndentedBuffer, Kernel
 from .triton import TritonKernel
 from .triton_utils import config_of, signature_to_meta
 
 
+@dataclass
+class PartitionState:
+    partitions: List[Tuple]
+    cur_partition: List[Tuple]
+    cur_count: int
+
+    def finalize(self):
+        if self.cur_partition:
+            self.partitions.append(self.cur_partition)
+
+
 class ForeachKernel(Kernel):
     MAX_NUM_ARGS = 250  # number where I would no longer get triton errors
 
     @staticmethod
-    def horizontal_partition(nodes):
-        """Generates a list of list of nodes where each node sublist is
-        guaranteed to not exceed CUDA limits for number of args (read/writes)."""
-        assert len(nodes) >= 1
-
-        cur_count = 0
-        partitions = []
-        cur_partition = []
-        for node in nodes:
+    def _update_partition(partition_state, node_rw_count, node_info):
+        if partition_state.cur_count + node_rw_count > ForeachKernel.MAX_NUM_ARGS:
+            partition_state.partitions.append(partition_state.cur_partition)
+            partition_state.cur_partition = [node_info]
+            partition_state.cur_count = node_rw_count
+        else:
+            partition_state.cur_count += node_rw_count
+            partition_state.cur_partition.append(node_info)
+
+    @staticmethod
+    def horizontal_partition(subkernel_nodes, triton_scheduling):
+        """Generates a list of lists of node info tuples which consist of (fused_nodes, tiling, numel, rnumel)
+        for each subkernel node where each sublist is guaranteed to not exceed CUDA limits for number of args
+        (read/writes) and to have the same 2D or 1D blocking strategy."""
+        assert len(subkernel_nodes) >= 1
+
+        partition_state_1d = PartitionState([], [], 0)
+        yelem_to_partition_state_2d = defaultdict(lambda: PartitionState([], [], 0))
+
+        for node in subkernel_nodes:
+            fused_nodes = node.get_nodes()
+            _, (numel, rnumel) = max(
+                fused_nodes, key=lambda x: int(x.is_reduction())
+            ).group
+            tiled_groups = triton_scheduling.select_tiling(fused_nodes, numel, rnumel)
+            node_info = fused_nodes, tiled_groups, numel, rnumel
+
             read_writes = node.read_writes
             read_write_count = len(read_writes.reads) + len(read_writes.writes)
-            if cur_count + read_write_count > ForeachKernel.MAX_NUM_ARGS:
-                partitions.append(cur_partition)
-                cur_partition = [node]
-                cur_count = read_write_count
+
+            if tiled_groups[1] == 1:
+                ForeachKernel._update_partition(
+                    partition_state_1d, read_write_count, node_info
+                )
             else:
-                cur_count += read_write_count
-                cur_partition.append(node)
+                y_elem = tiled_groups[0]
```

Comment:
So if the kernel is a reduction kernel, `tiled_groups[0]` will be numel and `tiled_groups[1]` will be rnumel. In this case `tiled_groups[0]` would be actually X-numel. Assigning it to `y_elem` looks a bit confusing here. Although I need understand more how we use `y_elem` in other places.

Reply:
I guess foreach kernel does not handle reduction? If that's the case, let's add an assertion for 'rnumel == 1' above

Reply:
Currently I don't lower any reduction kernels, but was planning to  in the future, how should I handle this consistently? or should I just add an if checking if it's a reduction?